contracts/SpokePoolV3Periphery.sol

```diff
@@ -350,93 +376,113 @@ contract SpokePoolV3Periphery is Lockable, MultiCaller {
-        uint8 v,
-        bytes32 r,
-        bytes32 s
+        bytes calldata permitSignature
     ) external nonReentrant {
+        (bytes32 r, bytes32 s, uint8 v) = _deserializeSignature(permitSignature);
         // For permit transactions, we wrap the call in a try/catch block so that the transaction will continue even if the call to
         // permit fails. For example, this may be useful if the permit signature, which can be redeemed by anyone, is executed by somebody
         // other than this contract.
-        try swapToken.permit(msg.sender, address(this), swapTokenAmount, deadline, v, r, s) {} catch {}
-        IERC20 _swapToken = IERC20(address(swapToken));
-        _swapToken.safeTransferFrom(msg.sender, address(this), swapTokenAmount);
-        _swapAndBridge(
-            exchange,
-            routerCalldata,
-            swapTokenAmount,
-            minExpectedInputTokenAmount,
-            depositData,
-            _swapToken,
-            acrossInputToken
+        try
+            IERC20Permit(address(swapData.swapToken)).permit(
+                msg.sender,
+                address(this),
+                swapData.swapTokenAmount,
+                deadline,
+                v,
+                r,
+                s
+            )
+        {} catch {}
+        IERC20 _swapToken = IERC20(address(swapData.swapToken));
+        _swapToken.safeTransferFrom(msg.sender, address(this), swapData.swapTokenAmount);
+        _swapAndBridge(swapData, depositData, acrossInputToken);
+    }
+
+    /**
+     * @notice Uses permit2 to transfer tokens from a user before swapping a token on this chain via specified router and submitting an Across deposit atomically.
+     * Caller can specify their slippage tolerance for the swap and Across deposit params.
+     * @dev This function assumes the caller has properly set an allowance for the permit2 contract on this network.
+     * @dev This function assumes that the amount of token to be swapped is equal to the amount of the token to be received from permit2.
+     * @param acrossInputToken Address of the token that will be bridged via Across as the inputToken.
+     * @param depositor The owner of the permit2 signature and depositor for the Across spoke pool.
+     * @param swapData Specifies the params we need to perform a swap on a generic exchange.
+     * @param depositData Specifies the Across deposit params we'll send after the swap.
+     * @param permit The permit data signed over by the owner.
+     * @param transferDetails The spender's requested transfer details for the permitted token.
+     * @param signature The permit2 signature to verify against the deposit data.
+     */
+    function swapAndBridgeWithPermit2(
+        IERC20 acrossInputToken,
+        address depositor,
+        SwapData calldata swapData,
+        DepositData calldata depositData,
+        IPermit2.PermitTransferFrom calldata permit,
+        IPermit2.SignatureTransferDetails calldata transferDetails,
```

Comment:
Can you not infer the transferDetails from the depositData or swapData?

Reply:
Good call. I checked with forge and it seems like we also save ~800 gas if we do this. https://github.com/across-protocol/contracts/pull/782/commits/e1162402065398e41221f13ffb2a25a39f2d03ef